### DISCUSSION TITLE
SSR support

### DISCUSSION THREAD

**Discussion by EchinopsisM** [codex-team/editor.js#1869] [Ideas]
> Can't use by node.js import in Next.js

When I try to use by way of a Node.js import, I get an "ReferenceError: window is not defined" error in the browser and can't use it at all.

It works fine when I load it by CDN in the head.

**Comment by danilockthar** [comment]
Hello Noah. Could you connect editor js to nextjs ? Im trying the cdn as you said but im not able to make it work. If you can give a little insight, maybe at this time you already make it work on next.

**Comment by findelallc** [comment]
@noahniuwa you need to wrap the editor as a component as make it ssr false through dynamic import method provided by Next.js. I got the same problem before.

you may follow like this
`const EditorWidget = dynamic(import ('../src/components/Utils/Editor'),{ssr:false});`

Now in the editor you can start adding configs and tool options as per your requirements.
Thanks.

**Comment by md-abdul-halim-rafi** [comment]
There is a similar problem in importing in the **GatsbyJs** project. It works fine in localhost with normal importing. But, at gatsby build it occurs an **ssr** issue. I solved it for my gatsby js project like this: 

import the EditorJs component by `@loadable/component`

```
import Loadable from "@loadable/component"
const EditorJs = Loadable(() => import("../../../../../editor/EditorJs"))
```

for more details, check gatsby js [official documentation](https://www.gatsbyjs.org/docs/using-client-side-only-packages)

**Comment by tkesgar** [comment]
I believe the problem is that the distributed editor.js package module (`dist/editor.js`) somehow tries to access `window`. This will cause issues with all SSRs (not only Next.js, but also Nuxt.js, Gatsby, etc.)

A simple fix is to import editor.js dynamically. Here is a simple wrapper we use for React:

```tsx
import * as React from "react";
import { EditorConfig } from "@editorjs/editorjs";

interface EditorJsWrapperProps extends React.ComponentProps<"div"> {
  config?: EditorConfig;
}

export default function EditorJsWrapper({
  config = {},
  ...restProps
}: EditorJsWrapperProps): JSX.Element {
  const elmtRef = React.useRef<HTMLDivElement>();

  React.useEffect(() => {
    if (!elmtRef.current) {
      return;
    }

    let editorJs;

    (async () => {
      const { default: EditorJS } = await import("@editorjs/editorjs");

      editorJs = new EditorJS({
        ...config,
        holder: elmtRef.current,
      });
    })().catch((error): void => console.error(error));

    return (): void => {
      editorJs.destroy();
    };
  }, [config]);

  return (
    <div
      {...restProps}
      ref={(elmt): void => {
        elmtRef.current = elmt;
      }}
    />
  );
}
```

A proper fix would be to check if editor.js **never access `window` on module level**.

**Comment by Teebo** [comment]
@neSpecc, @tkesgar mentions that we need a check on Editor.js itself if `window` is defined, I was searching in the source of editor.js but was unsure where the `window` object is being used, can you please point me to that file(s). 
Node does not give me the source map so I cannot point it out on the error log.

**Comment by Teebo** [comment]
@tkesgar I am using the Editor in Angular, I have the same issue when building for SSR.
I am just curious how does a dynamic import solves this issue. 
If you do not mind, please check this issue https://github.com/codex-team/editor.js/issues/1254

Because it fails even when the window object is under the globals;
I am just trying to make sense of this issue.

**Comment by tkesgar** [comment]
I don't remember the exact part of code that causes the issue, but I looked for `window` on the source code and these lines seem suspicious:

https://github.com/codex-team/editor.js/blob/33be2052edc18bef7c320dca3646be9a2aa954af/src/components/utils.ts#L166-L174

**Comment by b2m9** [comment]
Ran into this problem when building a statically rendered SvelteKit app. The build process fails because the package is trying to load `window` as soon as it is loaded.

**Comment by jbarrantesr92** [comment]
I am using Angular 11, and I get the window error during ssr only when define `const editor = new EditorJS ()`. If I put a comment block on this, then I don't get the window error even when the module is imported, and I can see my app in the server without the editor of course. I tried IsPlatformBrowser, but this does not work. I read about creating a shime module, but I'm not sure how to make it work for editorjs: https://medium.com/@caerus.karu/isplatformbrowser-is-not-your-friend-5ee52b06da89

**Comment by sisindrymedagam** [comment]
Hello,

I have integrated editor.js in my Angular SSR. when I was trying to load the editor route, the node server encountered an error saying `self is not defined`. Please help with it.

The complete log goes here...
```js
ERROR Error: Uncaught (in promise): ReferenceError: self is not defined
ReferenceError: self is not defined
    at Object.8941 (C:\home\site\wwwroot\vendors-node_modules_editorjs_code_dist_bundle_js-node_modules_editorjs_editorjs_dist_editor_-77895d.js:1:492565)
    at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
    at Object.9353 (C:\home\site\wwwroot\src_app_shared_fluent-ui_editorjs_editorjs_component_ts-src_app_shared_fluent-ui_editorjs_edi-493a3d.js:1:2836)
    at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
    at Module.2913 (C:\home\site\wwwroot\src_app_questions_qsingle_qsingle_module_ts.js:1:6433)
    at Function.__webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
    at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976150)
    at Object.onInvoke (C:\home\site\wwwroot\main.js:1:470317)
    at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976046)
    at Zone.run (C:\home\site\wwwroot\main.js:1:1968989)
    at resolvePromise (C:\home\site\wwwroot\main.js:1:1990849)
    at resolvePromise (C:\home\site\wwwroot\main.js:1:1989864)
    at C:\home\site\wwwroot\main.js:1:1992718
    at ZoneDelegate.invokeTask (C:\home\site\wwwroot\main.js:1:1976992)
    at Object.onInvokeTask (C:\home\site\wwwroot\main.js:1:469995)
    at ZoneDelegate.invokeTask (C:\home\site\wwwroot\main.js:1:1976882)
    at Zone.runTask (C:\home\site\wwwroot\main.js:1:1970002)
    at drainMicroTaskQueue (C:\home\site\wwwroot\main.js:1:1980199) {
  rejection: ReferenceError: self is not defined
      at Object.8941 (C:\home\site\wwwroot\vendors-node_modules_editorjs_code_dist_bundle_js-node_modules_editorjs_editorjs_dist_editor_-77895d.js:1:492565)
      at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
      at Object.9353 (C:\home\site\wwwroot\src_app_shared_fluent-ui_editorjs_editorjs_component_ts-src_app_shared_fluent-ui_editorjs_edi-493a3d.js:1:2836)
      at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
      at Module.2913 (C:\home\site\wwwroot\src_app_questions_qsingle_qsingle_module_ts.js:1:6433)
      at Function.__webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
      at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976150)
      at Object.onInvoke (C:\home\site\wwwroot\main.js:1:470317)
      at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976046)
      at Zone.run (C:\home\site\wwwroot\main.js:1:1968989),
  promise: ZoneAwarePromise [Promise] {
    __zone_symbol__state: 0,
    __zone_symbol__value: ReferenceError: self is not defined
        at Object.8941 (C:\home\site\wwwroot\vendors-node_modules_editorjs_code_dist_bundle_js-node_modules_editorjs_editorjs_dist_editor_-77895d.js:1:492565)
        at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
        at Object.9353 (C:\home\site\wwwroot\src_app_shared_fluent-ui_editorjs_editorjs_component_ts-src_app_shared_fluent-ui_editorjs_edi-493a3d.js:1:2836)
        at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
        at Module.2913 (C:\home\site\wwwroot\src_app_questions_qsingle_qsingle_module_ts.js:1:6433)
        at Function.__webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
        at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976150)
        at Object.onInvoke (C:\home\site\wwwroot\main.js:1:470317)
        at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976046)
        at Zone.run (C:\home\site\wwwroot\main.js:1:1968989)
  },
  zone: Zone {
    _parent: Zone {
      _parent: null,
      _name: '<root>',
      _properties: {},
      _zoneDelegate: [ZoneDelegate]
    },
    _name: 'angular',
    _properties: { isAngularZone: true },
    _zoneDelegate: ZoneDelegate {
      _taskCounts: [Object],
      zone: [Circular],
      _parentDelegate: [ZoneDelegate],
      _forkZS: null,
      _forkDlgt: null,
      _forkCurrZone: null,
      _interceptZS: null,
      _interceptDlgt: null,
      _interceptCurrZone: null,
      _invokeZS: [Object],
      _invokeDlgt: [ZoneDelegate],
      _invokeCurrZone: [Circular],
      _handleErrorZS: [Object],
      _handleErrorDlgt: [ZoneDelegate],
      _handleErrorCurrZone: [Circular],
      _scheduleTaskZS: [Object],
      _scheduleTaskDlgt: [ZoneDelegate],
      _scheduleTaskCurrZone: [Circular],
      _invokeTaskZS: [Object],
      _invokeTaskDlgt: [ZoneDelegate],
      _invokeTaskCurrZone: [Circular],
      _cancelTaskZS: [Object],
      _cancelTaskDlgt: [ZoneDelegate],
      _cancelTaskCurrZone: [Circular],
      _hasTaskZS: [Object],
      _hasTaskDlgt: [ZoneDelegate],
      _hasTaskDlgtOwner: [Circular],
      _hasTaskCurrZone: [Circular]
    }
  },
  task: ZoneTask {
    _zone: Zone {
      _parent: [Zone],
      _name: 'angular',
      _properties: [Object],
      _zoneDelegate: [ZoneDelegate]
    },
    runCount: 0,
    _zoneDelegates: null,
    _state: 'notScheduled',
    type: 'microTask',
    source: 'Promise.then',
    data: ZoneAwarePromise [Promise] {
      __zone_symbol__state: 0,
      __zone_symbol__value: ReferenceError: self is not defined
          at Object.8941 (C:\home\site\wwwroot\vendors-node_modules_editorjs_code_dist_bundle_js-node_modules_editorjs_editorjs_dist_editor_-77895d.js:1:492565)
          at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
          at Object.9353 (C:\home\site\wwwroot\src_app_shared_fluent-ui_editorjs_editorjs_component_ts-src_app_shared_fluent-ui_editorjs_edi-493a3d.js:1:2836)
          at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
          at Module.2913 (C:\home\site\wwwroot\src_app_questions_qsingle_qsingle_module_ts.js:1:6433)
          at Function.__webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
          at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976150)
          at Object.onInvoke (C:\home\site\wwwroot\main.js:1:470317)
          at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976046)
          at Zone.run (C:\home\site\wwwroot\main.js:1:1968989)
    },
    scheduleFn: undefined,
    cancelFn: undefined,
    callback: [Function],
    invoke: [Function]
  }
}
```

If there is a Solution, Please Suggest.

**Comment by tkesgar** [comment]
For those who still encounter this issue, the problem is that the package's code references browser-specific APIs when executed on server, which does not provide browser-specific APIs. This is an issue with most browser-intensive packages that interact and manipulate browser DOM, since adding a special check for SSR is just not worth it (leads to a lot of `if-then-else` and edge cases).

To work around the issue, you need to ensure that EditorJS code is not executed on SSR. You can refer to your framework documentation on how to achieve this:

- Next.js: use [`next/dynamic`](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr)
- Gatsby: use [`@loadable/component`](https://www.gatsbyjs.com/docs/using-client-side-only-packages/#workaround-3-load-client-side-dependent-components-with-loadable-components)

For example, here is how to import it in Next.js:

```jsx
// ComponentThatUseEditorJS.jsx
import EditorJS from "@editorjs/editor"

export default function ComponentThatUseEditorJS() {
  return /* your code here */
}

// MyPage.jsx
import dynamic from "next/dynamic"

const ComponentThatUseEditorJs = dynamic(() => import("./ComponentThatUseEditorJs"))

export default function MyPage() {
  return (
    <div>
      <ComponentThatUseEditorJs />
    </div>
  )
}
```

If you use third-party EditorJS wrappers such as [react-editor-js](https://github.com/Jungwoo-An/react-editor-js) or [vue-editor-js](https://github.com/ChangJoo-Park/vue-editor-js), the fix still applies: use your framework's tools to exclude it from SSR.

This technique also works with other non-SSR-compatible packages (not only EditorJS).

I propose to close this issue since the problem is not on EditorJS side and the workaround is platform-specific.

**Comment by Chaitanya-yash** [comment]
> Hello,
> 
> I have integrated editor.js in my Angular SSR. when I was trying to load the editor route, the node server encountered an error saying `self is not defined`. Please help with it.
> 
> The complete log goes here...
> 
> ```js
> ERROR Error: Uncaught (in promise): ReferenceError: self is not defined
> ReferenceError: self is not defined
>     at Object.8941 (C:\home\site\wwwroot\vendors-node_modules_editorjs_code_dist_bundle_js-node_modules_editorjs_editorjs_dist_editor_-77895d.js:1:492565)
>     at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>     at Object.9353 (C:\home\site\wwwroot\src_app_shared_fluent-ui_editorjs_editorjs_component_ts-src_app_shared_fluent-ui_editorjs_edi-493a3d.js:1:2836)
>     at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>     at Module.2913 (C:\home\site\wwwroot\src_app_questions_qsingle_qsingle_module_ts.js:1:6433)
>     at Function.__webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>     at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976150)
>     at Object.onInvoke (C:\home\site\wwwroot\main.js:1:470317)
>     at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976046)
>     at Zone.run (C:\home\site\wwwroot\main.js:1:1968989)
>     at resolvePromise (C:\home\site\wwwroot\main.js:1:1990849)
>     at resolvePromise (C:\home\site\wwwroot\main.js:1:1989864)
>     at C:\home\site\wwwroot\main.js:1:1992718
>     at ZoneDelegate.invokeTask (C:\home\site\wwwroot\main.js:1:1976992)
>     at Object.onInvokeTask (C:\home\site\wwwroot\main.js:1:469995)
>     at ZoneDelegate.invokeTask (C:\home\site\wwwroot\main.js:1:1976882)
>     at Zone.runTask (C:\home\site\wwwroot\main.js:1:1970002)
>     at drainMicroTaskQueue (C:\home\site\wwwroot\main.js:1:1980199) {
>   rejection: ReferenceError: self is not defined
>       at Object.8941 (C:\home\site\wwwroot\vendors-node_modules_editorjs_code_dist_bundle_js-node_modules_editorjs_editorjs_dist_editor_-77895d.js:1:492565)
>       at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>       at Object.9353 (C:\home\site\wwwroot\src_app_shared_fluent-ui_editorjs_editorjs_component_ts-src_app_shared_fluent-ui_editorjs_edi-493a3d.js:1:2836)
>       at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>       at Module.2913 (C:\home\site\wwwroot\src_app_questions_qsingle_qsingle_module_ts.js:1:6433)
>       at Function.__webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>       at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976150)
>       at Object.onInvoke (C:\home\site\wwwroot\main.js:1:470317)
>       at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976046)
>       at Zone.run (C:\home\site\wwwroot\main.js:1:1968989),
>   promise: ZoneAwarePromise [Promise] {
>     __zone_symbol__state: 0,
>     __zone_symbol__value: ReferenceError: self is not defined
>         at Object.8941 (C:\home\site\wwwroot\vendors-node_modules_editorjs_code_dist_bundle_js-node_modules_editorjs_editorjs_dist_editor_-77895d.js:1:492565)
>         at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>         at Object.9353 (C:\home\site\wwwroot\src_app_shared_fluent-ui_editorjs_editorjs_component_ts-src_app_shared_fluent-ui_editorjs_edi-493a3d.js:1:2836)
>         at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>         at Module.2913 (C:\home\site\wwwroot\src_app_questions_qsingle_qsingle_module_ts.js:1:6433)
>         at Function.__webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>         at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976150)
>         at Object.onInvoke (C:\home\site\wwwroot\main.js:1:470317)
>         at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976046)
>         at Zone.run (C:\home\site\wwwroot\main.js:1:1968989)
>   },
>   zone: Zone {
>     _parent: Zone {
>       _parent: null,
>       _name: '<root>',
>       _properties: {},
>       _zoneDelegate: [ZoneDelegate]
>     },
>     _name: 'angular',
>     _properties: { isAngularZone: true },
>     _zoneDelegate: ZoneDelegate {
>       _taskCounts: [Object],
>       zone: [Circular],
>       _parentDelegate: [ZoneDelegate],
>       _forkZS: null,
>       _forkDlgt: null,
>       _forkCurrZone: null,
>       _interceptZS: null,
>       _interceptDlgt: null,
>       _interceptCurrZone: null,
>       _invokeZS: [Object],
>       _invokeDlgt: [ZoneDelegate],
>       _invokeCurrZone: [Circular],
>       _handleErrorZS: [Object],
>       _handleErrorDlgt: [ZoneDelegate],
>       _handleErrorCurrZone: [Circular],
>       _scheduleTaskZS: [Object],
>       _scheduleTaskDlgt: [ZoneDelegate],
>       _scheduleTaskCurrZone: [Circular],
>       _invokeTaskZS: [Object],
>       _invokeTaskDlgt: [ZoneDelegate],
>       _invokeTaskCurrZone: [Circular],
>       _cancelTaskZS: [Object],
>       _cancelTaskDlgt: [ZoneDelegate],
>       _cancelTaskCurrZone: [Circular],
>       _hasTaskZS: [Object],
>       _hasTaskDlgt: [ZoneDelegate],
>       _hasTaskDlgtOwner: [Circular],
>       _hasTaskCurrZone: [Circular]
>     }
>   },
>   task: ZoneTask {
>     _zone: Zone {
>       _parent: [Zone],
>       _name: 'angular',
>       _properties: [Object],
>       _zoneDelegate: [ZoneDelegate]
>     },
>     runCount: 0,
>     _zoneDelegates: null,
>     _state: 'notScheduled',
>     type: 'microTask',
>     source: 'Promise.then',
>     data: ZoneAwarePromise [Promise] {
>       __zone_symbol__state: 0,
>       __zone_symbol__value: ReferenceError: self is not defined
>           at Object.8941 (C:\home\site\wwwroot\vendors-node_modules_editorjs_code_dist_bundle_js-node_modules_editorjs_editorjs_dist_editor_-77895d.js:1:492565)
>           at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>           at Object.9353 (C:\home\site\wwwroot\src_app_shared_fluent-ui_editorjs_editorjs_component_ts-src_app_shared_fluent-ui_editorjs_edi-493a3d.js:1:2836)
>           at __webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>           at Module.2913 (C:\home\site\wwwroot\src_app_questions_qsingle_qsingle_module_ts.js:1:6433)
>           at Function.__webpack_require__ (C:\home\site\wwwroot\main.js:1:2670977)
>           at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976150)
>           at Object.onInvoke (C:\home\site\wwwroot\main.js:1:470317)
>           at ZoneDelegate.invoke (C:\home\site\wwwroot\main.js:1:1976046)
>           at Zone.run (C:\home\site\wwwroot\main.js:1:1968989)
>     },
>     scheduleFn: undefined,
>     cancelFn: undefined,
>     callback: [Function],
>     invoke: [Function]
>   }
> }
> ```
> 
> If there is a Solution, Please Suggest.

+1
I have tried all possible soultions out there and unable to solve it still

**Comment by Teebo** [comment]
@Chaitanya-yash if you have domino installed, try

```
const template = readFileSync(join(distFolder, INDEX_HTML_PATH)).toString();
const win = domino.createWindow(template.toString());

global['self'] = win;
```

**Comment by Chaitanya-yash** [comment]
@Teebo Yeah man. I did try domino. Ediotr.js still throws an exception about the window. Had the same problem with Localstorage but that is gone now. This one has been haunting me for a couple of days now.

```
const domino = require('domino');
const template = readFileSync('dist/browser/index.html').toString();
const win = domino.createWindow(template.toString());

global['window'] = win;
global['localStorage'] = localStorage;
```

**Comment by Teebo** [comment]
@Chaitanya-yash so you've also tried?
Note the 'self' being defined

```
global['window'] = win;
global['self'] = window;
```

**Comment by Chaitanya-yash** [comment]
@Teebo Yeah, I am getting the window error right now upon multiple builds. Self is not coming up
`ReferenceError: window is not defined`

**Comment by Teebo** [comment]
@Chaitanya-yash can you share you `server.ts `code?

**Comment by Chaitanya-yash** [comment]
@Teebo 
```
import 'zone.js/dist/zone-node';

import { ngExpressEngine } from '@nguniversal/express-engine';
import * as express from 'express';
import { join } from 'path';

import { AppServerModule } from './src/main.server';
import { APP_BASE_HREF } from '@angular/common';
import { existsSync, readFileSync } from 'fs';
import 'localstorage-polyfill'

const distFolder = join(process.cwd(), 'dist/hbrdWeb/browser');

// const indexHtml = existsSync(join(distFolder, 'index.original.html')) ? 'index.original.html' : 'index';
const domino = require('domino');
const template = readFileSync('dist/browser/index.html').toString();
const win = domino.createWindow(template.toString());

global['window'] = win;
global['localStorage'] = localStorage;
global['self'] = window;

// The Express app is exported so that it can be used by serverless Functions.
export function app(): express.Express {
  const server = express();
  // const distFolder = join(process.cwd(), 'dist/hbrdWeb/browser');
  // const indexHtml = existsSync(join(distFolder, 'index.original.html')) ? 'index.original.html' : 'index';

  // Our Universal express-engine (found @ https://github.com/angular/universal/tree/master/modules/express-engine)
  server.engine('html', ngExpressEngine({
    bootstrap: AppServerModule,
  }));

  server.set('view engine', 'html');
  server.set('views', distFolder);

  server.get('*.*', express.static(distFolder, {
    maxAge: '1y'
  }));

  // All regular routes use the Universal engine
  server.get('*', (req, res) => {
    res.render(indexHtml, { req, providers: [{ provide: APP_BASE_HREF, useValue: req.baseUrl }] });
  });

  return server;
}

function run(): void {
  const port = process.env.PORT || 4000;

  // Start up the Node server
  const server = app();
  server.listen(port, () => {
    console.log(`Node Express server listening on http://localhost:${port}`);
  });
}

// Webpack will replace 'require' with '__webpack_require__'
// '__non_webpack_require__' is a proxy to Node 'require'
// The below code is to ensure that the server is run only when not requiring the bundle.
declare const __non_webpack_require__: NodeRequire;
const mainModule = __non_webpack_require__.main;
const moduleFilename = mainModule && mainModule.filename || '';
if (moduleFilename === __filename || moduleFilename.includes('iisnode')) {
  run();
}

export * from './src/main.server';
```

**Comment by Teebo** [comment]
@Chaitanya-yash let's start by trying to move the below

```
import { AppServerModule } from './src/main.server';
import { APP_BASE_HREF } from '@angular/common';
```

Move it underneath 

```
global['window'] = win;
global['localStorage'] = localStorage;
global['self'] = window;
```

**Comment by Teebo** [comment]
@Chaitanya-yash the below works for me, I talk it from somewhere online, I still need to make sense of it all. Because some lines use `win` some use `window`. But yeah, that works for me
```

const template = readFileSync(join(distFolder, indexHtml)).toString();
const win = domino.createWindow(template.toString());


global['window'] = win;
global['DOMTokenList'] = win.DOMTokenList;
global['Node'] = win.Node;
global['File'] = win.File;
global['Text'] = win.Text;
global['HTMLElement'] = win.HTMLElement;
global['navigator'] = win.navigator;
global['Event'] = win.Event;
global['Event']['prototype'] = win.Event.prototype;
global['window'] = window;
global['Element'] = window.Element;
global['self'] = window;
global['IDBIndex'] = window.IDBIndex;
global['document'] = window.document;
global['getComputedStyle'] = window.getComputedStyle;
global['MutationObserver'] = getMockMutationObserver();
function getMockMutationObserver() {
  return class {
    observe(node, options) { }

    disconnect() { }

    takeRecords() {
      return [];
    }
  };
}
Object.defineProperty(win.document.body.style, 'transform', {
  value: () => {
    return {
      enumerable: true,
      configurable: true
    };
  },
});
global['document'] = win.document;
global['CSS'] = null;

import { AppServerModule } from './src/main.server';
import { APP_BASE_HREF } from '@angular/common';
```

**Comment by Chaitanya-yash** [comment]
I spun up a new Angular universal project, added EditorJS to it, initialised it  `new EditorJS() `and I put everything on top in server.ts. It errored out giving the same error. @Teebo Thanks for helping out. Let me try something else.

**Comment by Chaitanya-yash** [comment]
@Teebo Can you share your the versions for Angular and EditorJS that you are using?

**Comment by b2m9** [comment]
Can you please use the discussion forums of this repo to sort this out? Thanks.

**Comment by EugeneGohh** [comment]
This is how I fixed it: 

`const Editor = dynamic(() => import("./Editor"), { ssr: false });`

**Comment by newlegendmedia** [comment]
> Ran into this problem when building a statically rendered SvelteKit app. The build process fails because the package is trying to load `window` as soon as it is loaded.

I am having the same issue. Did you find a solution for svelte-kit?

**Reply by robertcoopercode** [reply]
Got this working in svelte kit as follows:

```svelte
<script>
import { browser } from '$app/env';
import { onMount } from 'svelte';

onMount(async () => {
	if (browser) {
		const EditorJS = (await import('@editorjs/editorjs')).default;
		var editor = new EditorJS('editorjs');
	}
});
</script>

<div id="editorjs" />
```

**Comment by novikov1337danil** [comment]
Do I understand correctly that at the moment it is impossible to prepare rendered editorjs on the server side? 
I'm interested in displaying it in readOnly mode

**Reply by fluid-design-io** [reply]
@novikov1337danil I also wanted to display a rendered content for my Next.js app, and I created a simple custom `Renderer` that takes the `BlockData[]` and renders React components based on the type, here's what I have, feel free to add/modify them:

```tsx
// renderer.tsx
import { cn } from '@/lib/utils'
import { OutputData } from '@editorjs/editorjs'
import slugify from '@sindresorhus/slugify'

import { Separator } from '@ui/index'

const Header = ({ level, text }: { level: number; text: string }) => {
  const Tag = `h${level}` as keyof JSX.IntrinsicElements
  return (
    <Tag id={slugify(text)} className="scroll-m-20">
      {text}
    </Tag>
  )
}

const Paragraph = ({ text }: { text: string }) => {
// we need to do this because editorjs doesn't separate link tags
  return <p dangerouslySetInnerHTML={{ __html: text }}></p>
}

const List = ({ items, style }: { items: string[]; style: 'unordered' | 'ordered' }) => {
  const ListTag = style === 'unordered' ? 'ul' : 'ol'

  return (
    <ListTag>
      {items.map((item, index) => (
        <li key={index}>{item}</li>
      ))}
    </ListTag>
  )
}

const Image = ({
  url,
  caption,
  stretched,
  withBorder,
  withBackground,
}: {
  url: string
  caption: string
  stretched: boolean
  withBorder: boolean
  withBackground: boolean
}) => {
  return (
    <figure className="relative">
      <img
        src={url}
        alt={caption}
        className={cn(
          'object-contain',
          { 'w-full': stretched },
          { 'rounded-md': !stretched },
          { border: withBorder },
          { 'bg-muted': withBackground }
        )}
        sizes="(max-width: 768px) 100vw, 33vw"
      />
      {caption && <figcaption>{caption}</figcaption>}
    </figure>
  )
}

const Renderer = ({ blocks }: { blocks: OutputData['blocks'] }) => {
  return (
    <div>
      {blocks.map((block) => {
        switch (block.type) {
          case 'header':
            return <Header key={block.id} level={block.data.level} text={block.data.text} />
          case 'paragraph':
            return <Paragraph key={block.id} text={block.data.text} />
          case 'list':
            return <List key={block.id} items={block.data.items} style={block.data.style} />
          case 'divider':
            return <Separator />
          case 'image':
            return (
              <Image
                key={block.id}
                url={block.data.file.url}
                caption={block.data.caption}
                stretched={block.data.stretched}
                withBorder={block.data.withBorder}
                withBackground={block.data.withBackground}
              />
            )
          default:
            return (
              <p className="text-destructive" key={block.id}>
                Unsupported block type: {block.type}
              </p>
            )
        }
      })}
    </div>
  )
}

export default Renderer
```
Here's how I use it
```tsx
// blog/[slug]/page.tsx
import Renderer from '@/components/renderer'

import { cn } from '@/lib/utils'

import { OutputData } from '@editorjs/editorjs'


type BlogProps = {
  params: { slug: string }
}

export default async function BlogPage({ params }: BlogProps) {
const { slug } = params
// your fetch logic to get the editor contents
const blog = await getBlog(slug)
// assume your editorjs data is stored inside blog.body
const body = blog.body as any
// get the blocks
const blocks: OutputData['blocks'] = body.blocks

return (
  <div className="prose dark:prose-invert">
            <Renderer blocks={blocks} />
  </div>
)
}
```

**Comment by bekogeko** [comment]
Im feeling the same, I feels like it will be good idea to create a
editorjs-nextjs compatible library for this.

There are options but none of them work properly for v14, only few of them
work(sometimes) in v13/12
